raw_tiles/index/index.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import absolute_import
+import gzip
+from msgpack import Unpacker
+from raw_tiles.tile import Tile
+from io import BufferedReader
+from raw_tiles.index.features import FeatureTileIndex
+from raw_tiles.index.highway import HighwayIndex
+from raw_tiles.index.routes import RouteIndex
+
+
+def tile_contents(tile, table):
+    """
+    Generator yielding each item in a gzipped msgpack format file.
+
+    TODO: This should be generalised? Perhaps move into formatter classes?
+    """
+
+    file_name = 'tiles/osm/%s/%d/%d/%d.msgpack.gz' % \
+        (table, tile.z, tile.x, tile.y)
+
+    with BufferedReader(gzip.open(file_name, 'rb')) as gz:
+        unpacker = Unpacker(file_like=gz)
+        for obj in unpacker:
+            yield obj
+
+
+def index_table(source, index_fn, *indices):
+    """
+    Index a table for the given tile coordinates. The `index_fn` is called
+    with each item in the tile for each of `*indices` which has that function
+    defined.
+    """
+
+    # filter only indices which respond to index_fn
+    indexable = list()
+    for index in indices:
+        if callable(getattr(index, index_fn, None)):
+            indexable.append(index)
```

Comment:
I personally find this more reflective style for organization harder to piece together (checking for callables with a particular name to opt into behavior), and prefer when there's a more explicit abstraction or structure behind it. But, that's just a matter of personal taste really. (Also not sure if things just emerged this way organically, or if you set out to organize it this way from the get go :)

Reply:
I was trying to make sure that the code only iterates over each table once, to avoid repeated overheads of deserialisation. But since the object types are different for each table, that led to this reflection-based approach.

The function is probably simple enough that we can just have it 3 times; `index_feature_table`, `index_ways_table`, `index_rels_table`, if you think that would be more readable?

I'm slightly worried that we know we'll be adding WOF and the land / water polygon tables soon, and a bunch of heterogenous NE tables at some point after that (only for z<10?). All of which might mean a profusion of different object functions.

Do you think it might work better to have a single function taking the `obj` tuple, and leave it to the receiving function to unpack and assert that the input is the right "shape" of row for the table?

Reply:
> The function is probably simple enough that we can just have it 3 times; index_feature_table, index_ways_table, index_rels_table, if you think that would be more readable?

Probably? :) I was just thinking of grouping the indexes beforehand more explicitly. In other words, we'd know beforehand which indexes and how to create them, applied to which table, and what function would get called. Right now the `index` function knows a little bit about that, because it knows the function names to call on the indices.

```
def index(get_table, *indices):
    planet_osm_ways = get_table('planet_osm_ways')
    index_table(planet_osm_ways, 'add_way', *indices)

    planet_osm_rels = get_table('planet_osm_rels')
    index_table(planet_osm_rels, 'add_relation', *indices)
    for typ in ('point', 'line', 'polygon'):
        source = get_table('planet_osm_' + typ)
        index_table(source, 'add_feature', *indices)
```

I'm just proposing to take that a little bit further and have the indices grouped already on the way in, so we wouldn't use reflection to figure that out. Something more along the lines of:

```
def index(get_table, way_indices, rel_indices, feature_indices):
    planet_osm_ways = get_table('planet_osm_ways')
    index_table(planet_osm_ways, way_indices)

    planet_osm_rels = get_table('planet_osm_rels')
    index_table(planet_osm_rels, rel_indices)

    for typ in ('point', 'line', 'polygon'):
        source = get_table('planet_osm_' + typ)
        index_table(source, feature_indices)
```

Longer term I can see this list of tables and indices mapping be driven more from configuration. Maybe that would help alleviate some of your concerns about all the other tables that need to be added? Although I would expect most of it to just use the feature index, no?

Anyway, I could be missing something a piece of the puzzle here as to why this might be a bad idea. This was meant to be more of a knitpick rather than anything else!

Reply:
Thanks! I had a go at this in 7f97144. Does that look better?

Reply:
Yup, looks great to me!